### ISSUE TITLE
Questions for generating sequence using ProtXLNet pertained-model

### ISSUE THREAD

**Issue by gundalav** [agemagician/ProtTrans#108]
Dear Sir,

I hope this message finds you well. First of all, I would like to express my admiration for the excellent work you have done with ProtTrans. Your contributions to the field of protein language modeling are truly remarkable.

I have a few questions regarding the implementation you shared on [ipynb document for Sequence Generation](https://github.com/agemagician/ProtTrans/blob/master/Generate/ProtXLNet.ipynb). Specifically, I would like to inquire about the following:

I am curious to know if the input sequence strictly needs to be separated by white spaces. For instance, if we remove the white space in a sequence (e.g., "A E T C Z A O"), will it affect the results?

Is it possible to introduce masks in the input sequence? For example, using a sequence like `A E T [MASK] [MASK] [MASK] A O` where the inference is made only on the masked area for 3 residues.

Additionally, I was wondering if there are any known use cases or publications regarding sequence generation using ProtXLnet.

Thank you for your time and expertise. Your guidance and insights would be greatly appreciated.

Best regards,
G.V.

**Comment by mheinzinger** [comment]
Thanks for your kind words! :)

Yes, all our models/tokenizers need to get input separated by white-space. 
Only this way the model/tokenizer recognizes single amino acids as single tokens. If you were to pass two amino acids not separated by white-space, the model will not know which token to map this to and map it to the unknown token automatically.
You can test this by yourself by simply checking the tokens produced by the tokenizer when passing a sequence separated by white-space and one without white-spaces.

Yes, you can introduce [MASK] tokens in the input and ask the model to reconstruct them in the output. I would recommend to use ProtT5 for this. In this case, you will have to introduce T5-specific mask tokens (<extra_id_0>, <extra_id_1> etc. See [huggingface example](https://huggingface.co/docs/transformers/model_doc/t5#inference) for this). We also use this under-the-hood for ProtT5-based variant effect prediction called [VESPA](https://github.com/Rostlab/VESPA#log-odds-ratio-output). This gives you automatically the log-odds for each position, i.e.,  we mask one position/residue at a time, ask the model to reconstruct it and compute log-odds from the model outputs. If you want to get mask probabilities for multiple residues at a time, you might need to slightly modify our [VESPA masking strategy](https://github.com/Rostlab/VESPA/blob/main/vespa/predict/logodds.py#L122). There you see that we only replace one residue at a time with <extra_id_0> and reconstruct this residue. If you were to mask multiple residues at a time, you would have to increase the counter of the <extra_id_X>. So if you were to mask two residues, you would have to give the model sth like `P R T <extra_id_0> I N S E Q W <extra_id_1> E N C E`

**Comment by gundalav** [comment]
Dear Michael,

Thank you so much for your reply. I tried your suggestions with this code:

```
from transformers import AutoTokenizer, AutoModel
from transformers import T5EncoderModel, T5Tokenizer, T5ForConditionalGeneration
import torch
import re

tokenizer = T5Tokenizer.from_pretrained('Rostlab/prot_t5_xl_half_uniref50-enc', do_lower_case=False)
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
model = T5ForConditionalGeneration.from_pretrained("Rostlab/prot_t5_xl_half_uniref50-enc")
model = model.to(device)
model = model.eval()


sequence = "P R T <extra_id_0> I N S E Q W <extra_id_1> E N C E"
# this will replace all rare/ambiguous amino acids by X and introduce white-space between all amino acids
sequence = re.sub(r"[UZOB]", "<unk>", sequence)

# tokenize sequences and pad up to the longest sequence in the batch
input_ids = tokenizer(sequence, return_tensors="pt").input_ids.to(device)
sequence_ids = model.generate(input_ids)
sequences = tokenizer.batch_decode(sequence_ids)
print(sequences)
```

It gives me this output.

```
['<pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad>']
```

Is this expected? If no, what should I do to get a more reasonable result?

Thanks and hope to hear from you again.


G.V.

**Comment by mheinzinger** [comment]
Hm, I have to admit that I did not experiment with masking multiple residues at a time so far.
So I can only give you some more or less educated guess on your problem:
- The current example sequence is relatively short. Given that always only 15% of the input tokens got masked during training, your example might be already out-of-distribution. So maybe give it a short with more realistic sequence lengths.
- the .generate function takes an argument which allows you to generate multiple answers (e.g. do_sample=True). Check this tutorial for details: https://huggingface.co/blog/how-to-generate . This way you can generate multiple answers and pick the ones that have no preliminary end. You might be able to experiment with different .generate settings and judge their quality by providing labels to the model-call as done during training: https://huggingface.co/docs/transformers/model_doc/t5#training

**Comment by gundalav** [comment]
> Hm, I have to admit that I did not experiment with masking multiple residues at a time so far (at least not during inference). So I can only give you some more or less educated guess on your problem:
> 
> * The current example sequence is relatively short. Given that always only 15% of the input tokens got masked during training, your example might be already out-of-distribution. So maybe give it a shot with more realistic sequence lengths.
> * the .generate function takes an argument which allows you to generate multiple answers (e.g. do_sample=True). Check this tutorial for details: https://huggingface.co/blog/how-to-generate . This way you can generate multiple answers and pick the ones that have no preliminary end. You might be able to experiment with different .generate settings and judge their quality by providing labels to the model-call as done during training: https://huggingface.co/docs/transformers/model_doc/t5#training



Hi Michel,
I tried with a longer sequence length. And it works. Thanks to your suggestion.


```
sequence =  "P R T <extra_id_0> I N S E Q W <extra_id_1> <extra_id_2> N C E I M K R L W K K"
```

Sincerely,
G.V.
